client/client.go

```diff
@@ -247,26 +214,170 @@ func (c *Client) update(latestRoot bool) (data.TargetFiles, error) {
+	// download intermediate root metadata files, until the
+	// latest available one is reached. Therefore, it MUST
+	// temporarily turn on consistent snapshots in order to
+	// download versioned root metadata files as described next.
+
+	// This loop returns on error or breaks after downloading the lastest root metadata.
+	// 5.3.2 Let N denote the version number of the trusted root metadata file.
+	for i := 0; i < c.UpdaterMaxRoots; i++ {
+		// 5.3.3 Try downloading version nPlusOne of the root metadata file.
+		// NOTE: as a side effect, we do update c.rootVer to nPlusOne between iterations.
+		nPlusOne := c.rootVer + 1
+		nPlusOneRootPath := util.VersionedPath("root.json", nPlusOne)
+		nPlusOneRootMetadata, err := c.downloadMetaUnsafe(nPlusOneRootPath, defaultRootDownloadLimit)
+
+		if err != nil {
+			if _, ok := err.(ErrMissingRemoteMetadata); ok {
+				// stop when the next root can't be downloaded
+				break
+			} else {
+				return err
+			}
+		}
+
+		// 5.3.4 Check for an arbitrary software attack.
+		nPlusOneRootMetadataSigned := &data.Root{}
+		if nPlusOne == 1 {
+			return ErrNoRootKeys
+		} else {
+			// 5.3.4.1 Check that N signed N+1
+			// 5.3.5 Check for a rollback attack. Here, we check that nPlusOneRootMetadataSigned.version >= nPlusOne.
+			if err := c.db.UnmarshalExpired(nPlusOneRootMetadata, nPlusOneRootMetadataSigned, "root", nPlusOne); err != nil {
+				// 5.3.6 Note that the expiration of the new (intermediate) root
+				// metadata file does not matter yet, because we will check for
+				// it in step 5.3.10.
+				if _, ok := err.(verify.ErrExpired); !ok {
+					return err
+				}
+			}
+		}
+
+		// 5.3.5 Following up, we check for a fast-forward attack: here, we check for that nPlusOneRootMetadataSigned.version >= nPlusOne.
+		if nPlusOneRootMetadataSigned.Version != nPlusOne {
+			return verify.ErrWrongVersion{
+				Given:    nPlusOneRootMetadataSigned.Version,
+				Expected: nPlusOne,
+			}
+		}
+
+		// 5.3.7 Set the trusted root metadata file to the new root metadata file.
+		c.rootVer = nPlusOneRootMetadataSigned.Version
+		// NOTE: following up on 5.3.1, we want to always have consistent snapshots on for the duration
+		// of root rotation. AFTER the rotation is over, we will set it to the value of the last root.
+		consistentSnapshot = nPlusOneRootMetadataSigned.ConsistentSnapshot
+		// 5.3.8 Persist root metadata. The client MUST write the file to non-volatile storage as FILENAME.EXT (e.g. root.json).
+		// NOTE: Internally, setMeta stores metadata in LevelDB in a persistent manner.
+		if err := c.local.SetMeta("root.json", nPlusOneRootMetadata); err != nil {
+			return err
+		}
+
+		for k := range startKeyIDs {
+			if !reflect.DeepEqual(startKeyIDs[k], getKeyIDs(k)) {
+				c.local.SetMeta(k, json.RawMessage{})
+			}
+		}
+
+		// 5.3.4.2 check that N+1 signed itself.
+		//This is different from the previous call because now the threashold and the keys are updated.
```

Comment:
nit: threshold typo

is it really that they have been updated, or that `getLocalRootMeta()` verifies using the updated keys (by creating a NewDB https://github.com/theupdateframework/go-tuf/blob/31841328f398d1235ef6da054162985747f000eb/client/client.go#L450)

Reply:
The intent is to verify them. However, `getLocalRootMeta` comes with the side effect of removing all the keys by creating a NewDB and loading new keys.

Reply:
Unfortunately, the current implementation does not comply with what you described. In order for the getLocalRootMeta to work, we store the 'root.json' permanently by calling `c.local.SetMeta("root.json", ...)` and then do one extra check (N+1 signed itself). Therefore, in case of failure, it can not be recovered next time (@trishankatdatadog can you please double check).

The issue at the moment is first line `meta, err := c.local.GetMeta()` in the `getLocalRootMeta` (as well as the previous implementation of the getLocalMeta) that loads metadata from levelDB. 

I am up for a change if we agree on more changes here. It does not make sense to leave the client in a bad state. WDYT?